### DISCUSSION TITLE
What tsconfig settings does lit recommend?

### DISCUSSION THREAD

**Discussion by cadamsdev** [lit/lit#3923] [Q&A] [answered]
Is the documentation on this page outdated?
https://lit.dev/docs/tools/publishing/#compiling-with-typescript

I'm looking at the lit-element-starter-ts repo and it's using slightly different settings
https://github.com/lit/lit-element-starter-ts/blob/main/tsconfig.json

Questions
For "target" is there a reason it's using "es2019"? We should be able to bump that up to at least "es2020"+ by now since it's widely supported
https://caniuse.com/?search=es2020
https://kangax.github.io/compat-table/es2016plus/

For "module" should we use "es2015" or "es2020" or does it not really matter.

For "lib" should I use "DOM.Iterable"?

**Comment by usergenic** [comment]
Apart from what we are recommending strongly in https://lit.dev/docs/tools/publishing/#compiling-with-typescript the other tsconfig concerns are mostly left up to the developer to address the specifics of their audience, pipeline, and deployment scenario etc.

AFAICT the primary reason for using newer targets is to reduce the runtime cost and bloat of inlined compilation of newer features to older analogs.  Advancing the module support to es2020 from es2015 shouldn't pose an issue as far as I'm aware and I believe the recommendations on our page are meant to define a comfortable lower-bound where there is little to no syntactical "downsampling" in target compilation.

**Reply by augustjk** [reply, marked as answer]
To add to this:

We are making a small update to the doc site for v3 as we do plan on bumping our target to `es2021`. https://lit.dev/docs/v3/tools/publishing/#compiling-with-typescript

`"module"` doesn't matter much beyond `es2015` unless you're doing dynamic imports, use `import.meta` or do top level awaits. https://www.typescriptlang.org/tsconfig#es2015es6es2020es2022

`"DOM.Iterable"` is needed if you need DOM lists like `NodeList` to be iterable, like to `for...of` or spread, etc. We do need it in our library code as we do take advantage of that, but I want to say if you don't need it yourself, it might be okay to drop it.

**Reply by justinfagnani** [reply]
I think we should probably update that section to point out which settings are basically required, which are `experimentalDecorators: true`, `useDefineForClassFields: true`, and a `target` of ES6 or later (to avoid needing the custom elements shim).

Nothing else is really a blocker that I know of. You could even compile to AMD or CommonJS if you're bundling to include Lit.

**Reply by cadamsdev** [reply]
Thanks everyone, very informative 🙂 that helps a lot.
